browser_patches/firefox/juggler/NetworkObserver.js

```diff
@@ -180,6 +187,77 @@ class NetworkRequest {
     this._interceptedChannel = undefined;
   }
 
+  async interceptResponse(url, method, headers, postData) {
+    const uri = url ? Services.io.newURI(url) : this.httpChannel.URI;
+    const newChannel = NetUtil.newChannel({
+      uri,
+      loadingNode: this.httpChannel.loadInfo.loadingContext,
+      loadingPrincipal: this.httpChannel.loadInfo.loadingPrincipal || this._interceptedChannel.loadInfo.principalToInherit,
+      triggeringPrincipal: this.httpChannel.loadInfo.triggeringPrincipal,
+      securityFlags: this.httpChannel.loadInfo.securityFlags,
+      contentPolicyType: this.httpChannel.loadInfo.internalContentPolicyType,
+    }).QueryInterface(Ci.nsIRequest).QueryInterface(Ci.nsIHttpChannel);
+    newChannel.loadInfo = this.httpChannel.loadInfo;
+    newChannel.loadGroup = this.httpChannel.loadGroup;
+
+    for (const header of (headers || requestHeaders(this.httpChannel)))
+      newChannel.setRequestHeader(header.name, header.value, false /* merge */);
+
+    if (postData) {
+      setPostData(newChannel, postData, headers);
+    } else if (this.httpChannel instanceof Ci.nsIUploadChannel) {
+      newChannel.QueryInterface(Ci.nsIUploadChannel);
+      newChannel.setUploadStream(this.httpChannel.uploadStream, '', -1);
+    }
+    // We must set this after setting the upload stream, otherwise it
+    // will always be 'PUT'. (from another place in the source base)
+    newChannel.requestMethod = method || this.httpChannel.requestMethod;
+
+    this._networkObserver._responseInterceptionChannels.add(newChannel);
+    const body = await new Promise((resolve, reject) => {
+      NetUtil.asyncFetch(newChannel, (stream, status) => {
+        this._networkObserver._responseInterceptionChannels.delete(newChannel);
+        if (!Components.isSuccessCode(status)) {
+          reject(status);
+          return;
+        }
+        try {
+          resolve(NetUtil.readInputStreamToString(stream, stream.available()));
+        } catch (e) {
+          if (e.result == Cr.NS_BASE_STREAM_CLOSED) {
+            // The stream was empty.
+            resolve('');
+          } else {
+            reject(e);
+          }
+        } finally {
+          stream.close();
+        }
+      });
+    });
+
+    const pageNetwork = this._activePageNetwork();
+    if (pageNetwork)
+      pageNetwork._responseStorage.addResponseBody(this, newChannel, body);
+
+    const responseHeaders = [];
```

Comment:
Do you want a helper method for lines 243-256? I think we have a copy of that :)

Reply:
Yeah, added.